From 3d2c6d683e51a913db7f4c1e964dd0bc3fd4eab9 Mon Sep 17 00:00:00 2001
From: Peter Fang <peter.fang@intel.com>
Date: Thu, 2 Mar 2023 02:11:59 -0800
Subject: [PATCH 5/8] OvmfPkg: Workarounds for TDP

Due to the lack of support for TD HOB and TdxDxe in TDP, work around
them for now and fix them later.

Signed-off-by: Peter Fang <peter.fang@intel.com>
---
 .../BaseMemEncryptTdxLib/MemoryEncryption.c   | 20 ++++++++-----------
 OvmfPkg/PlatformPei/MemDetect.c               |  2 +-
 OvmfPkg/TdxDxe/TdxDxe.c                       |  2 +-
 3 files changed, 10 insertions(+), 14 deletions(-)

diff --git a/OvmfPkg/Library/BaseMemEncryptTdxLib/MemoryEncryption.c b/OvmfPkg/Library/BaseMemEncryptTdxLib/MemoryEncryption.c
index a71b1efbca..620a679e05 100644
--- a/OvmfPkg/Library/BaseMemEncryptTdxLib/MemoryEncryption.c
+++ b/OvmfPkg/Library/BaseMemEncryptTdxLib/MemoryEncryption.c
@@ -587,21 +587,17 @@ SetOrClearSharedBit (
   }
 
   //
-  // If changing shared to private, must accept-page again
+  // If changing shared to private, must accept-page again if needed
   //
   if (Mode == ClearSharedBit) {
     Status = gBS->LocateProtocol (&gEdkiiMemoryAcceptProtocolGuid, NULL, (VOID **)&MemoryAcceptProtocol);
-    if (EFI_ERROR (Status)) {
-      DEBUG ((DEBUG_ERROR, "%a: Failed to locate MemoryAcceptProtocol with %r\n", __func__, Status));
-      ASSERT (FALSE);
-      return Status;
-    }
-
-    Status = MemoryAcceptProtocol->AcceptMemory (MemoryAcceptProtocol, PhysicalAddress, Length);
-    if (EFI_ERROR (Status)) {
-      DEBUG ((DEBUG_ERROR, "%a: Failed to AcceptMemory with %r\n", __func__, Status));
-      ASSERT (FALSE);
-      return Status;
+    if (!EFI_ERROR (Status)) {
+      Status = MemoryAcceptProtocol->AcceptMemory (MemoryAcceptProtocol, PhysicalAddress, Length);
+      if (EFI_ERROR (Status)) {
+        DEBUG ((DEBUG_ERROR, "%a: Failed to AcceptMemory with %r\n", __func__, Status));
+        ASSERT (FALSE);
+        return Status;
+      }
     }
   }
 
diff --git a/OvmfPkg/PlatformPei/MemDetect.c b/OvmfPkg/PlatformPei/MemDetect.c
index 493cb1fbeb..518a2693f9 100644
--- a/OvmfPkg/PlatformPei/MemDetect.c
+++ b/OvmfPkg/PlatformPei/MemDetect.c
@@ -349,7 +349,7 @@ InitializeRamRegions (
   IN EFI_HOB_PLATFORM_INFO  *PlatformInfoHob
   )
 {
-  if (TdIsEnabled ()) {
+  if (TdIsEnabled () && !TdpIsEnabled ()) {
     PlatformTdxPublishRamRegions ();
     return;
   }
diff --git a/OvmfPkg/TdxDxe/TdxDxe.c b/OvmfPkg/TdxDxe/TdxDxe.c
index 30732f421b..f150d08cd6 100644
--- a/OvmfPkg/TdxDxe/TdxDxe.c
+++ b/OvmfPkg/TdxDxe/TdxDxe.c
@@ -341,7 +341,7 @@ TdxDxeEntryPoint (
   SetPcdSettings (PlatformInfo);
  #endif
 
-  if (!TdIsEnabled ()) {
+  if (!TdIsEnabled () || TdpIsEnabled ()) {
     //
     // If it is Non-Td guest, we install gEfiMpInitLibMpDepProtocolGuid so that
     // MpInitLib will be used in CpuDxe driver.
-- 
2.43.0

